File: PhaseII/backend/src/models/task.py
from sqlmodel import SQLModel, Field, Relationship
from typing import Optional
from datetime import datetime
import uuid
from enum import Enum


class TaskStatus(str, Enum):
    pending = "pending"
    completed = "completed"


class TaskBase(SQLModel):
    title: str = Field(min_length=1, max_length=200, sa_column_kwargs={"index": True})
    description: Optional[str] = Field(default=None, max_length=1000)
    status: TaskStatus = Field(default=TaskStatus.pending)
    due_date: Optional[datetime] = Field(default=None)


class Task(TaskBase, table=True):
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    user_id: uuid.UUID = Field(foreign_key="user.id", ondelete="CASCADE", sa_column_kwargs={"index": True})
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    # Relationship to User
    user: Optional["User"] = Relationship(back_populates="tasks")


class TaskCreate(TaskBase):
    pass


class TaskRead(TaskBase):
    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class TaskUpdate(SQLModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    due_date: Optional[datetime] = None

File: PhaseIV/shared/types/task.ts
export type TaskStatus = 'pending' | 'completed';

export interface Task {
  id: string;
  title: string;
  description?: string;
  status: TaskStatus;
  user_id: string;
  created_at: string; // ISO date string
  updated_at: string; // ISO date string
  due_date?: string; // ISO date string (optional)
}

export interface TaskCreateRequest {
  title: string;
  description?: string;
  due_date?: string; // ISO date string (optional)
}

export interface TaskUpdateRequest {
  title?: string;
  description?: string;
  due_date?: string; // ISO date string (optional)
}

export interface TaskToggleCompleteRequest {
  completed: boolean;
}

export interface TaskListResponse {
  tasks: Task[];
}